server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java

```diff
@@ -141,25 +145,27 @@ private ContextIndexSearcher(
         MutableQueryTimeout cancellable,
         int minimumDocsPerSlice,
         boolean wrapWithExitableDirectoryReader,
-        ThreadPoolExecutor executor
+        BoundedExecutor executor,
+        boolean forceSequentialCollection
     ) throws IOException {
         // we need to pass the executor up so it can potentially be used as a sliceExecutor by knn search
         super(wrapWithExitableDirectoryReader ? new ExitableDirectoryReader((DirectoryReader) reader, cancellable) : reader, executor);
         setSimilarity(similarity);
         setQueryCache(queryCache);
         setQueryCachingPolicy(queryCachingPolicy);
         this.cancellable = cancellable;
-        this.queueSizeBasedExecutor = executor != null ? new QueueSizeBasedExecutor(executor) : null;
         this.minimumDocsPerSlice = minimumDocsPerSlice;
-        if (executor != null) {
-            this.leafSlices = computeSlices(
-                getLeafContexts(),
-                queueSizeBasedExecutor.threadPoolExecutor.getMaximumPoolSize(),
-                minimumDocsPerSlice
-            );
-            assert (this.leafSlices.length <= executor.getMaximumPoolSize());
-        } else {
+        if (executor == null) {
             this.leafSlices = null;
+        } else {
+            if (forceSequentialCollection) {
+                // we offload to the executor unconditionally, including requests that don't support concurrency
+                // TODO are we good offloading aggs that don't support concurrency to the concurrent executor or is postCollect an issue?
```

Comment:
yes that aspect is covered but postCollect at times does the entire collection and that needs to happen on the same thread where the collection itself happens. I lost track of where we are with that, and whether offloading may cause problems or not. Hopefully we'll see test failures if so, as this PR does not enable concurrency for the query phase but it does enable the offloading to the separate thread pool for sequential execution.

Reply:
To share more details, the problem used to be assertions in Lucene that verified that doc_values read as part of post collect were being read in the same thread where the collection happens. We said the aggs that require this may not support concurrency to start with, yet they may run into issues even when running them sequentially only because we offload them to the separate pool. To be verified.

Reply:
@martijnvg do you have thoughts on this? Is it ok to always offload the collection even if sequential (single slice) including for requests that include aggs that explicitly don't support concurrency?

Reply:
Martijn and I chatted and spotted three aggs that are problematic, for which we are disabling offloading sequential collection (nested, composite and cardinality). There have been IT tests failures around these in this PR, which is good. A good explanation of the different aggs that are problematic when it comes to inter-segment concurrency is at #96023 and it clearly explains the two category of problems, out of which one involves not only parallel collection but also sequential collection on a separate thread.

As a follow-up we will see whether we can move postCollect to the worker thread and come up with a long term plan for these.

Reply:
postCollect has been moved to the search workers, and this is no longer a problem.